bitcoin/src/address.rs

```diff
@@ -565,22 +621,55 @@ impl<'a> fmt::Display for AddressEncoding<'a> {
 /// * [BIP341 - Taproot: SegWit version 1 spending rules](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki)
 /// * [BIP350 - Bech32m format for v1+ witness addresses](https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki)
 #[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
```

Comment:
One annoying property of these derives is they require `V` to implement the traits. We should just implement them manually.

Unrelated but I think deriving `PartialOrd` was wrong. I would expect the addresses to be lexicographically sorted but this sorts according to the internal representation.

Reply:
Why was this marked as resolved if the derives are still here?

Reply:
Sorry, I did not understand you would like them implemented manually as part of this PR. Would you be OK if it's done as follow-up? It is not completely related to this change and I can imagine it might spark new set of discussions. I could look at it right after this PR.

Reply:
I believe it's strongly related to this PR because that's literally doing this PR right. But since the downsides are mostly about documentation being bloated/less readable and a bunch of more code the compiler must process I think it's not too bad to put it in a separate PR if it helps you.

FTR AFAIK nobody is working in this file so you shouldn't get git conflicts anytime soon if you decide to do it in this one.

Reply:
Got it. I prefer opening new PR for it. It might take me some time to figure out some of the implementations.

Reply:
The `derivative` crate can actually help but we have strict policy on dependencies here. @apoelstra what do you think? It has MSRV 1.36, doesn't look maintained but presumably does what's needed so maybe that's not an issue (esp. for proc macro). I don't know the author but I plan to review the crate anyway for my project and I would publish `crev` proof.

Reply:
One thing I realized. If we remove the derives from the marker types (`NetworkChecked`, `NetworkUnchecked`), I think the following would not be possible:

```rust
#[derive(Clone)]
pub struct MyStruct<V: NetworkValidation> {
    pub address: Address<V>
}

#[test]
fn test_my_struct() {
    let s: MyStruct<NetworkChecked> = todo!();
    let s2 = s.clone();
}
```
This compiles when `NetworkChecked` has `Clone`, otherwise it does not.

So although rust-bitcoin library may not need the derives (because we could manually implement all the traits directly on `Address`), people may want to use the marker types in their own types that use derivation.

---

At this moment, similar problem exists with `Debug` because `Address<V>` does not implement `Debug` and I guess the compiler does not know that all `NetworkValidation` implementations have it (they don't have it in this PR, but I added it in my experiments).

In this situation, the compiler suggests to add `where Address<V>: Debug` to `MyStruct` or `#[derive(Debug)]` to `Address<V>`. I haven't figured out yet how to deal with this situation.

Reply:
Hmm, yes, I was thinking that we would deal with the annoyance so downstream users don't have to but maybe it's not that simple. But making the struct generic like you showed already has other problems, so I'm not sure if we should be very worried about it. Most people probably should have the field fixed.